### PR TITLE
[plugin/aws][feat] Add support for Lambda Functions

### PR DESCRIPTION
**Pull request by fernandocarletti**
# Description
Add support for Lambda Functions

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://resoto.com/code-of-conduct).

### CODE REVIEW

**Comment by lloesche** [comment]
@fernandocarletti thank you for the PR! Code looks great. 🥳 
One question and one comment I had while reading the code:

Question:
`role`, `runtime`, `code_size`, etc. are all typed `Optional`, but during collection the code says e.g.
```
lambda_function.runtime = function["Runtime"]
``` 
Is `runtime` really `Optional`? If yes then it should be `lambda_function.runtime = function.get("Runtime")`. Or, if you know that the boto3 API will always return a `Runtime` then the `runtime` attribute on the class shouldn't be `Optional`.

Also you could pass all those attributes directly to the constructor instead of setting them "later". 

So like
```
            lambda_function = AWSLambdaFunction(
                arn,
                tags,
                account=self.account,
                region=region,
                name=name,
                arn=arn,
                runtime=function["Runtime"],
                code_size=int(function["CodeSize"]),
                memory_size=function["MemorySize"],
                mtime=function["LastModified"],
                role=function["Role"],
                kms_key_arn=function.get("KmsKeyArn"),
              )

```

Again thank you for the PR! 👍

**Comment by fernandocarletti** [comment]
They are not optional. TBH I had it in the back of my mind to revisit it but ended up forgetting about it. I'll make those adjustments.

I was trying to find a way to unit test those changes, but I could not find an example of that. Can you point me to any example, please?

**Comment by lloesche** [comment]
@fernandocarletti re unit tests, skip them for now. We're currently rewriting the AWS collector. If you check the tests directory you'll see a bunch of unit tests for all kinds of boto3 methods. The old collector was not written in a way that would allow for easy mocking of the remote responses. Once merged we'll port your lambda collection.

**Comment by lloesche** [comment]
@fernandocarletti do you feel this is ready to merge?

File: resotolib/resotolib/baseresources.py [hunk 1]
```diff
@@ -854,6 +854,12 @@ class BaseBucket(BaseResource):
     kind: ClassVar[str] = "bucket"
 
 
+@define(eq=False, slots=False)
+class BaseServerlessFunction(BaseResource):
+    kind: ClassVar[str] = "serverless_function"
+    fingerprint: str = ""
```
Comment (aquamatthias) [inline resotolib/resotolib/baseresources.py:860]: Just stumble upon this property, which is not filled in the lambda collector. Why is this?

File: plugins/aws/resoto_plugin_aws/accountcollector.py [hunk 2]
```diff
@@ -1244,6 +1249,62 @@ def collect_instances(self, region: AWSRegion, graph: Graph) -> None:
...
+                    graph.add_edge(subnet, lambda_function)
+                    graph.add_edge(subnet, lambda_function, edge_type=EdgeType.delete)
+
+            for security_group_id in vpc_config.get("SecurityGroupIds", {}):
+                security_group = graph.search_first("id", security_group_id)
+                if security_group:
+                    graph.add_edge(security_group, lambda_function)
+                    graph.add_edge(security_group, lambda_function, edge_type=EdgeType.delete)
+
+            # TODO get ENIs
```
Comment (aquamatthias) [inline plugins/aws/resoto_plugin_aws/accountcollector.py:1304]: Is this a left over? Where would the ENIs come from?

**Comment by fernandocarletti** [comment]
@lloesche I still have some changes to apply here. I have a weird type mismatch in one of the fields that I need to fix. I'm having some busy weeks, I could not get back to it, but I plan to finish It by the weekend.

If you guys are in a hurry, feel free to finish it :)

File: plugins/aws/resoto_plugin_aws/accountcollector.py (hunk 2 above)
Comment (fernandocarletti) [reply plugins/aws/resoto_plugin_aws/accountcollector.py]: That was a note for me. Sadly it seems there's no clean way to get them. Although the IPs get assigned the only clear relation I could find in the API was the description of the ENI, which didn't seem like a scalable way to query it.

**Comment by fernandocarletti** [comment]
@lloesche I just fixed some conflicts and fixed the issues mentioned in the review. It should be ready. Thanks for your patience!

**Comment by fernandocarletti** [comment]
Actually, never mind. I just noticed it was superseded by #1022. I'm closing this one!

**Comment by lloesche** [comment]
@fernandocarletti it'll be a bit before #1022 will be active. My best estimate: around 2 weeks. I'm fine with merging your code into the existing collector. Code looks good to me! 👍 

Like I mentioned, we're refactoring the entire AWS collector, but before we activate the new code a series of extensive tests need to be run in different user's real world environments, to make sure it's on par with the existing collector code and doesn't break anything. If you're interested in that, #1023 is the draft PR that has the new code activated.

The existing code is used in production environments so we're very careful about not disturbing those, even on `main`. Esp. the cleanup functionality is crucial and before we activate such a massive code update that essentially replaces the entire collector we'll do lots of manual tests for every single resource.

The new collector code has the advantage that all resources return 100% of the information that the AWS API returns. And also adding new AWS services has become much much easier. Plus full test coverage.

**Review by lloesche — APPROVED** [review]

**Pull request merged**